src/shogun/base/SGObject.h

```diff
@@ -567,6 +567,20 @@ class CSGObject
 			demangled_type<T>().c_str());
 		return nullptr;
 	}
+
+	template <class T> const T* as() const
+	{
+		auto c = dynamic_cast<const T*>(this);
+		if (c)
+			return c;
+
+		SG_SERROR(
+			"Object of type %s cannot be converted to type %s.\n",
+			demangled_type<std::remove_pointer_t<decltype(this)>>().c_str(),
```

Comment:
btw can we re-use code from the non-const version in here? Like one calls the other?